rfcs/20230622-quantized-reduction.md

```diff
@@ -0,0 +1,605 @@
+a `result` tensor.
+
+More formally:
+...
+
+* `result[result_index] = reduce([source_values], [init_value], [0],
+        input_conversion, scatter, output_conversion)`
+ where:
+ ...
+
+#### Inputs
+
+| Label | Name                | Type     |
+|-------|---------------------|----------|
+| (I8)  | `input_conversion`  | function |
+| (I8)  | `scatter`           | function |
+| (I8)  | `output_conversion` | function |
+
+#### Constraints
+
+<!-- markdownlint-disable line-length -->
+* (C1) `element_type(operand) = element_type(source)`.
+* (C3) `element_type(init_value) = element_type(operand)`.
+* (C?) `baseline_element_type(inputs...) = baseline_element_type(results...)`.
+* (C?) `input_conversion` has type `tensor<E> -> (tensor<E'>)` where
+       `Ei = element_type(operand)`.
+* (C10) `scatter` has type `(tensor<E>, tensor<E>) -> tensor<E>` where
+        `E = element_type(output_types(input_conversion))`.
+* (C?) `output_conversion` has type `tensor<E> -> (tensor<E'>)` where
+       `E'i = element_type(result)`.
+* (C?) `element_type(output_types(input_conversion)) =
+       element_type(input_types(output_conversion))`.
+* (C11) `shape(operand) = shape(result)`.
+<!-- markdownlint-enable line-length -->
+
+## [11 Aug'23] Revised proposal
+
+### Context
+
+Option #2 should be avoided because it is hard to control the transformation
+which might disrupt the pattern to be matched. The option #1 sounds good except
+that the extra input/output conversion blocks are surplus information. The
+specification would benefit if the intent of the conversion blocks can be
+expressed precisely. The conversion blocks provides a way to capture the
+accumulation type needed to compute the accumulative operation on.
+
+The revised proposal is:
+
+* To capture the accumulation type via an additional StableHLO attribute like
+  `accumulation_element_type`.
+* The attribute seems beneficial for other ops as well like `dot_general` and
+  `convolution`.
+* `precision_config`, currently used for `dot_general` and `convolution`, is
+  used to override the precision specified by the input parameters, allowing the
+  choice of low precision vs high precision computation. We should consider
+  adding `precision_config` to all reduction based op as well.
+
+### Few implementation details
+
+#### On StableHLO side
+
+The reduce syntax to be augmented with a optional [type
+attribute](https://github.com/llvm/llvm-project/blob/51a57074bc63842970c4c160b05c1a7e42db7523/mlir/include/mlir/IR/OpBase.td#L1466)
+as follows:
+
+```mlir
+%0 = stablehlo.reduce(%arg0 init: %arg1) across dimensions = [0] {
+    accumulation_type = tensor<!quant.uniform<i32:f32, 3.400000e+01:16>>
+} : (tensor<16x!quant.uniform<i8:f32, 3.400000e+01:16>>, tensor<!quant.uniform<i8:f32, 3.400000e+01:16>>) -> tensor<!quant.uniform<i8:f32, 3.400000e+01:16>>
+    reducer(%arg2: tensor<!quant.uniform<i32:f32, 3.400000e+01:16>>, %arg3: tensor<!quant.uniform<i32:f32, 3.400000e+01:16>>)  {
```

Comment:
IMO, the revised plan doesn't work.  Unlike `ConvolutionOp` or `DotGeneralOp` which only uses multiply-accumulation in their computation, `ReduceOp` and `ReduceWindowOp` allow users to specify the `computation`.  For example, the computation can be `ADD`, `ARGMAX`, etc.  In the ARGMAX example, we may mix `QUI8` (values) and `I32` (indices) in the `computation` region.  What does it mean when we specify `accumulation_type` attribute as proposed?

Second, what does it mean when user specify `accumulation_type` in this particular sum-reduction example?  Apparently, if we compile the code as-is, we will unnecessary round/truncate values when we compute 8-bit `stablehlo.add`.  It will be unfortunate if we require the backend to guess which part must be replaced by `accumulation_type` and which part must not (or even worse, making this representation unusable).

If the concern is about that `input_conversion` and `output_conversion` regions are too flexible, I would rather specify that:

> If (1) the input operand type is different from the `computation` block argument type or (2) the op result type is different from the `computation` block return type, there will be implicit type conversion defined by either `stablehlo.convert`, `stablehlo.uniform_quantize`, or `stablehlo.uniform_dequantize`.

Reply:
Hi Logan
Thanks for the feedback!

>> It will be unfortunate if we require the backend to guess which part must be replaced by accumulation_type and which part must not (or even worse, making this representation unusable).

The reducer block  
`reducer(%arg2: tensor<!quant.uniform<i32:f32, 3.400000e+01:16>>, %arg3: tensor<!quant.uniform<i32:f32, 3.400000e+01:16>>)` is using the accumulation-typed block arguments to do computations and will be different from the input value type. The type of the attribute `accumulation_type`  must match the block arguments type and block return type.

IMO, the attribute is used to uniformly treat all the reduction based operations (including convolution and dot_general) to explicitly specify a accumulation type. For convolution and dot_general as there is no explicut reduction block, the attribute will inform the backend about the type to accumulate on. For reduction operation, the backend can just use the attribute rather than doing an equality type comparison between block argument type and input  value type and using the block argument type if the comparison is false. 

Let me know what do you think.

Reply:
How does the `accumulation_type` attribute work if there are more than one operations in the `reduce_computation` block?  

How does the backend rewrite the `reduce_computation` block if they want to use `accumulation_type`?

For example, if someone writes a fancy sum-max-argmax reduction and wants to accumulate in `QI32` but keep other stuff unchanged, how do they write it?  How does the backend interpret or compile it?

```
%init_sum = "stablehlo.constant"() {
    value = dense<input_zp> : tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>}
    : () -> tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>
%init_max_value = "stablehlo.constant"() {
    value = dense<MIN_VALUE> : tensor<...>}
    : () -> tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>
%init_max_idx = "stablehlo.constant"() {
    value = dense<0> : tensor<i32>} : () -> tensor<i32>
%indices = "stablehlo.iota"() {dim = 0} : () -> tensor<... x i32>

%result:3 = "stablehlo.reduce"(%input, %input, %indices, %init_sum, %init_max_value, %init_max_idx) ({
    ^reduce(%sum_lhs, %max_lhs, %max_idx_lhs, %sum_rhs, %max_rhs, %max_idx_rhs):
        // sum-reduction
        %sum_value = "stablehlo.add"(%sum_lhs, %sum_rhs)
            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>,
               tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>)
            -> tensor<!quant.uniform<ui8:f32, output_scale:output_zp>>

        // max-reduction and argmax-reduction
        %cmp = "stablehlo.compare"(%max_lhs, %max_rhs) {GE}
            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>,
               tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>)
            -> tensor<i1>
        %max_value = "stablehlo.select"(%cmp, %max_lhs, %max_rhs)
            : (...)
            -> tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>
        %eq = "stablehlo.compare"(%max_lhs, %max_rhs) {EQ}
            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>,
               tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>)
            -> tensor<i1>
        %first_idx = "stablehlo.minimum"(%max_idx_lhs, %max_idx_rhs)
            : (tensor<i32>, tensor<i32>) -> tensor<i32>>
        %max_idx = "stablehlo.select"(%cmp, %max_idx_lhs, %max_idx_rhs)
            : (tensor<i1>, tensor<i32>, tensor<i32>) -> tensor<i32>
        %new_idx = "stablehlo.select"(%eq, %first_idx, %max_idx)
            : (tensor<i1>, tensor<i32>, tensor<i32>) -> tensor<i32>
        "stablehlo.return"(%sum_value, %max_value, %new_idx)
            : (tensor<!quant.uniform<ui8:f32, output_scale:output_zp>>,
               tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>,
               tensor<i32>) -> ()
    }) {
      dimensions = dense<...> : tensor<1xi64>,
      accumulation_type = !quant.uniform<i32:f32, input_scale:0>
    } : (tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>,
         tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>,
         tensor<... x i32>,
         tensor<i32>)
    -> (tensor<... x !quant.uniform<ui8:f32, output_scale:output_zp>>,
        tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>,
        tensor<... x i32>)
```

Reply:
Thanks for the elaborate example. That is helpful!

In the above code (A) the type of block arguments `sum_lhs` and `sum_rhs` has to be explicitly expressed as QI32 and (B) `sum_value` needs to be computed at QI32, and (C) the block return include the `sum_value` with QI32 type. I can see that not all block arguments need to be considered with higher accumulation type (like `max_lhs, max_idx_lhs, max_rhs, max_idx_rhs`) and agree that the attribute is not sufficient to convey such granular information.

Let me think a bit more on this this and get back.